ruby/lib/gherkin/query.rb

```diff
@@ -20,34 +27,42 @@ def update_feature(feature)
       store_nodes_location(feature.tags)
 
       feature.children.each do |child|
-        update_rule(child.rule) if child.rule
-        update_background(child.background) if child.background
-        update_scenario(child.scenario) if child.scenario
+        update_rule(feature, child.rule) if child.rule
+        update_background(feature, child.background) if child.background
+        update_scenario(feature, child.rule, child.scenario) if child.scenario
       end
     end
 
-    def update_rule(rule)
+    def update_rule(feature, rule)
       return if rule.nil?
       store_nodes_location(rule.tags)
-
       rule.children.each do |child|
-        update_background(child.background) if child.background
-        update_scenario(child.scenario) if child.scenario
+        update_background(rule, child.background) if child.background
+        update_scenario(feature, rule, child.scenario) if child.scenario
       end
     end
 
-    def update_background(background)
+    def update_background(parent, background)
       update_steps(background.steps)
+      parent_id = parent.respond_to?(:id) ? parent.id : parent.object_id # Cucumber::Messages::Feature doesn't have an id :(
```

Comment:
Then I'm assuming this might need tweaking?

Reply:
IMO, the "correct" thing to do here would be to add a `Cucumber::Messages::Feature#id` method, to make it consistent with the rest of the message types, and the we could get rid of this workaround. It seems like an obvious hole to me. But it also seemed out of scope for this PR.

It seems to me that we have some options:
1. I can push another commit adding `C::M::F#id`
2. I can do a follow-up PR cleaning this up

I'm leaning towards 2 to avoid scope creep, but you're the boss! What do you think?

Reply:
@luke-hill Oh derp, 1 is impossible because its a different repo.

Reply:
@luke-hill I guess option 1 would be opening a PR on https://github.com/cucumber/messages , and waiting for that to be merged and released, and then we could merge and release this with a bumped requirement on that version, and then finally merge and release the cucumber-ruby-core PR with a bumped requirement on this.

Reply:
@luke-hill Just checked with crossed fingers, but `C::M::Feature` is still missing `#id` on current master : https://github.com/cucumber/messages/blob/main/ruby/lib/cucumber/messages.deserializers.rb#L244

Reply:
@luke-hill Ugh, and it looks like that code is all auto-generated. Adding an id to feature would change the data structure for ALL implementations of cucumber, not just the ruby one.

Maybe the simplest method would be to just use `#object_id` as the internal data structure key for both Feature and Rule. Or maybe just the object itself!

Reply:
@luke-hill Okay, I just pushed a commit that uses the object itself as the key instead, which ends up simplifying everything greatly, including removing the need for so many safe navigation operators. I'm happy with this if you are!